akka-projection-kafka/src/test/scala/akka/projection/kafka/integration/KafkaToSlickIntegrationSpec.scala

```diff
@@ -204,18 +204,58 @@ class KafkaToSlickIntegrationSpec extends KafkaSpecBase(ConfigFactory.load().wit
         assertAllOffsetsObserved(projectionId, topicName)
       }
     }
+
+    // relates to issue https://github.com/akka/akka-projection/issues/305
+    // we must ensure that we can read a kafka offset from the storage
+    "recover a projection from a previously saved offset" in {
+      val topicName = createTopic(suffix = 0, partitions = 3, replication = 1)
+      val groupId = createGroupId()
+      val projectionId = ProjectionId(groupId, "UserEventCountProjection-1")
+
+      produceEvents(topicName)
+
+      def slickProjection() = {
+
+        val kafkaSourceProvider: SourceProvider[MergeableOffset[JLong], ConsumerRecord[String, String]] =
+          KafkaSourceProvider(system.toTyped, consumerDefaults().withGroupId(groupId), Set(topicName))
```

Comment:
I discovered a bug in the KafkaSourceProvider. 

The `MetadataClientAdapter` can only be used once and that's problematic. 

The adapter has an actor that terminates when the projection stops. On restart we get the following exceptions: 

```scala
[info]   akka.pattern.AskTimeoutException: Recipient [Actor[akka://Spec/system/kafkaSourceProviderConsumer-3#780230987]] had already been terminated. Message of type [akka.kafka.Metadata$GetPartitionsFor].
[info]   at akka.pattern.AskableActorRef$.recipientTerminatedException(AskSupport.scala:293)
[info]   at akka.pattern.AskableActorRef$.internalAsk$extension(AskSupport.scala:338)
[info]   at akka.pattern.AskableActorRef$.$qmark$extension(AskSupport.scala:330)
[info]   at akka.kafka.scaladsl.MetadataClient.getPartitionsFor(MetadataClient.scala:75)
[info]   at akka.projection.kafka.internal.MetadataClientAdapterImpl.$anonfun$numPartitions$1(MetadataClientAdapter.scala:60)
[info]   at scala.collection.StrictOptimizedIterableOps.map(StrictOptimizedIterableOps.scala:99)
[info]   at scala.collection.StrictOptimizedIterableOps.map$(StrictOptimizedIterableOps.scala:86)
[info]   at scala.collection.immutable.Set$Set1.map(Set.scala:156)
[info]   at akka.projection.kafka.internal.MetadataClientAdapterImpl.numPartitions(MetadataClientAdapter.scala:60)
[info]   at akka.projection.kafka.internal.KafkaSourceProviderImpl.source(KafkaSourceProviderImpl.scala:86)
[info]   at akka.projection.internal.InternalProjectionState.$anonfun$mappedSource$1(InternalProjectionState.scala:374)
[info]   at scala.concurrent.impl.Promise$Transformation.run(Promise.scala:433)
[info]   at akka.dispatch.BatchingExecutor$AbstractBatch.processBatch(BatchingExecutor.scala:56)
[info]   at akka.dispatch.BatchingExecutor$BlockableBatch.$anonfun$run$1(BatchingExecutor.scala:93)
[info]   at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.scala:18)
[info]   at scala.concurrent.BlockContext$.withBlockContext(BlockContext.scala:94)
[info]   at akka.dispatch.BatchingExecutor$BlockableBatch.run(BatchingExecutor.scala:93)
[info]   at akka.dispatch.TaskInvocation.run(AbstractDispatcher.scala:48)
```

To ensure we get a new actor on each restart, I'm wrapping everything on this `def slickProjection` method, but that's not the way to go. Obviously. 

I will open another issue to address this.

Reply:
:+1:

Reply:
Does this test case also cover the bug you fixed in #324?

Reply:
Yes, #324 is the fix for that.

Reply:
Thanks @renatocaval , I stumbled upon this one as well. Thanks for providing fixes so quickly!